### DISCUSSION TITLE
Do I need a migration when going from TEXT --> DATETIME (both ISO strings)

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#2630] [Q&A] [answered]
Hi! I've just migrated my Database model in Drift so it will store datetimes correctly, using the build options to store them as ISO strings. Previously, I had a TEXT column already storing them as such ISO strings.

From my application code, I now pass Dart `DateTime` objects to the database, but will I actually need a database migration too? Since under the hood, all TEXT columns are still just TEXT columns.

**Comment by simolus3** [comment, marked as answer]
If you've been using `DateTime.toIso8601String()` before, a migration is needed for non-UTC date time values in the database.

To represent UTC offsets, we're following the format preferred by sqlite3 [here](https://sqlite.org/lang_datefunc.html#time_values), so we're directly appending the value (e.g. `+01:00`). This is different to `DateTime.toIso8601String()`, which does not add any suffix at all. Since sqlite3 datetime functions all use UTC internally and don't return a suffix either, we assume that date times without this suffix are in UTC.

So if a user is in Europe/Berlin and the (zone-local) value `DateTime(2023, 09, 23, 11, 29, 30)` was stored as `2023-09-23T11:29:30` in the database, drift would interpret the value as `DateTime.utc(2023, 09, 23, 11, 29, 30)` afterwards. That time represents a different instant though, it's two hours behind.
I think the easiest way to migrate those values is to convert them to UTC. I haven't tried it out, but I think something like this should to the trick:

```dart
Future<void> migrateFromTextToDateTime(Migrator m) async {
  for (final table in allTables) {
    final dateTimeColumns =
        table.$columns.where((c) => c.type == DriftSqlType.dateTime);

    if (dateTimeColumns.isNotEmpty) {
      // This table has dateTime columns which need to be migrated.
      await m.alterTable(TableMigration(
        table,
        columnTransformer: {
          for (final column in dateTimeColumns)
            // If there is a `Z` suffix already, it's UTC and everything is
            // good. Otherwise, explicitly convert to local time!
            column: CaseWhenExpression(
              cases: [
                // Already in UTC?
                CaseWhen(column.dartCast<String>().like('%Z'), then: column),
              ],
              // Alright, it's a local date time. Convert to UTC.
              orElse:
                  column.dartCast<DateTime>().modify(DateTimeModifier.utc()),
            ),
        },
      ));
    }
  }
}
```

**Reply by frankvollebregt** [reply]
Awesome, thank you. In my case, all datetimes happen to already be stored in UTC (with the Z appended), so no other steps should be required :)
